# Patch to override dokuwiki's page link generation routines when using Farm.
#
# It's a workaround for backwards compatibility when migrating to DokuWiki Farm,
# where we automatically inject a root-level namespace with the same name as the
# animal ID (e.g., `courses/uso/` injects `uso:` to all requests so it emulates
# the old link structure);
#
# Since we now derive the root namespace from the sub-wiki name, we must tell
# DokuWiki to automatically strip it from all generated URLs (e.g., to prevent 
# rendering duplicate namespaces, e.g.: `/courses/uso/uso/...`).
# Fortunately, it was pretty straight-forward, only needing to modify a single
# function generating URLs and some of its callers.

diff --git i/inc/common.php w/inc/common.php
index a2f3b3102..82a992f0b 100644
--- i/inc/common.php
+++ w/inc/common.php
@@ -457,11 +457,12 @@ function breadcrumbs()
  *
  * @param string $id pageid being filtered
  * @param bool $ue apply urlencoding?
+ * @param bool $rewrite apply custom namespace rewrite for farm
  * @return string
  * @author Andreas Gohr <andi@splitbrain.org>
  *
  */
-function idfilter($id, $ue = true)
+function idfilter($id, $ue = true, $rewrite = false)
 {
     global $conf;
     /* @var Input $INPUT */
@@ -469,6 +470,14 @@ function idfilter($id, $ue = true)
 
     $id = (string)$id;
 
+    if (defined("DOKU_FARM_ANIMAL") && !empty(DOKU_FARM_ANIMAL) && $rewrite) {
+        $prefix = DOKU_FARM_ANIMAL;
+        if (substr($id, 0, strlen($prefix)) == $prefix) {
+            $id = substr($id, strlen($prefix));
+            if (($id[0] ?? '') == ':') {$id = substr($id, 1);}
+        }
+    }
+
     if ($conf['useslash'] && $conf['userewrite']) {
         $id = strtr($id, ':', '/');
     } elseif (
@@ -515,7 +524,7 @@ function wl($id = '', $urlParameters = '', $absolute = false, $separator = '&amp
     if ($id === '') {
         $id = $conf['start'];
     }
-    $id = idfilter($id);
+    $id = idfilter($id, true, true);
     if ($absolute) {
         $xlink = DOKU_URL;
     } else {
@@ -562,7 +571,7 @@ function exportlink($id = '', $format = 'raw', $urlParameters = '', $abs = false
     }
 
     $format = rawurlencode($format);
-    $id = idfilter($id);
+    $id = idfilter($id, true, true);
     if ($abs) {
         $xlink = DOKU_URL;
     } else {
@@ -649,7 +658,7 @@ function ml($id = '', $more = '', $direct = true, $sep = '&amp;', $abs = false)
         return $xlink;
     }
 
-    $id = idfilter($id);
+    $id = idfilter($id, true, false);
 
     // decide on scriptname
     if ($direct) {
diff --git i/inc/html.php w/inc/html.php
index eedd2afe0..3140fa221 100644
--- i/inc/html.php
+++ w/inc/html.php
@@ -210,7 +210,7 @@ function html_btn($name, $id, $akey, $params, $method = 'get', $tooltip = '', $l
         $label = $lang['btn_' . $name];
 
     //filter id (without urlencoding)
-    $id = idfilter($id, false);
+    $id = idfilter($id, false, $conf['userewrite']);
 
     //make nice URLs even for buttons
     if ($conf['userewrite'] == 2) {
